dbms/src/Storages/DeltaMerge/Remote/RNWorkerFetchPages.cpp

```diff
@@ -0,0 +1,276 @@
+// limitations under the License.
+
+#include <Common/TiFlashMetrics.h>
+#include <Flash/Mpp/GRPCCompletionQueuePool.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/SharedContexts/Disagg.h>
+#include <Storages/DeltaMerge/Remote/ObjectId.h>
+#include <Storages/DeltaMerge/Remote/RNLocalPageCache.h>
+#include <Storages/DeltaMerge/Remote/RNReadTask.h>
+#include <Storages/DeltaMerge/Remote/RNWorkerFetchPages.h>
+#include <common/logger_useful.h>
+#include <kvproto/disaggregated.pb.h>
+#include <kvproto/kvrpcpb.pb.h>
+
+using namespace std::chrono_literals;
+
+namespace pingcap::kv
+{
+
+template <>
+struct RpcTypeTraits<disaggregated::FetchDisaggPagesRequest>
+{
+    using RequestType = disaggregated::FetchDisaggPagesRequest;
+    using ResultType = disaggregated::PagesPacket;
+    static std::unique_ptr<grpc::ClientReader<disaggregated::PagesPacket>> doRPCCall(
+        grpc::ClientContext * context,
+        std::shared_ptr<KvConnClient> client,
+        const RequestType & req)
+    {
+        return client->stub->FetchDisaggPages(context, req);
+    }
+    static std::unique_ptr<grpc::ClientAsyncReader<disaggregated::PagesPacket>> doAsyncRPCCall(
+        grpc::ClientContext * context,
+        std::shared_ptr<KvConnClient> client,
+        const RequestType & req,
+        grpc::CompletionQueue & cq,
+        void * call)
+    {
+        return client->stub->AsyncFetchDisaggPages(context, req, &cq, call);
+    }
+};
+
+} // namespace pingcap::kv
+
+namespace DB::DM::Remote
+{
+
+RNLocalPageCache::OccupySpaceResult blockingOccupySpaceForTask(const RNReadSegmentTaskPtr & seg_task)
+{
+    std::vector<PageOID> cf_tiny_oids;
+    {
+        cf_tiny_oids.reserve(seg_task->meta.delta_tinycf_page_ids.size());
+        for (const auto & page_id : seg_task->meta.delta_tinycf_page_ids)
+        {
+            auto page_oid = PageOID{
+                .store_id = seg_task->meta.store_id,
+                .ks_table_id = {seg_task->meta.keyspace_id, seg_task->meta.physical_table_id},
+                .page_id = page_id,
+            };
+            cf_tiny_oids.emplace_back(page_oid);
+        }
+    }
+
+    // Note: We must occupySpace segment by segment, because we need to read
+    // at least the complete data of one segment in order to drive everything forward.
+    // Currently we call occupySpace for each FetchPagesRequest, which is fine,
+    // because we send one request each seg_task. If we want to split
+    // FetchPagesRequest into multiples in future, then we need to change
+    // the moment of calling `occupySpace`.
+    auto page_cache = seg_task->meta.dm_context->db_context.getSharedContextDisagg()->rn_page_cache;
+
+    Stopwatch w_occupy;
+    auto occupy_result = page_cache->occupySpace(cf_tiny_oids, seg_task->meta.delta_tinycf_page_sizes);
+    // This metric is per-segment.
+    GET_METRIC(tiflash_disaggregated_breakdown_duration_seconds, type_cache_occupy).Observe(w_occupy.elapsedSeconds());
```

Comment:
A part of `tiflash_disaggregated_breakdown_duration_seconds` is per-segment while a part of it is per-request.
Splitting them into two metrics and Grafana panels will be clearer than mixing them into one.

Reply:
I will try to refine it in later PR